### PR TITLE
Implement SDL-0108 - AutoComplete List

### LINKED ISSUE #790: [SDL 0108] AutoCompleteList
Proposal: [AutoCompleteList](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0108-AutoCompleteList.md)

> The AutoCompleteList will give apps the ability to have multiple listing options while searching for things like destination and information; instead of giving a single suggestion.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/331

### PR DESCRIPTION
**Pull request by joeljfischer**
Fixes #790 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Unit tests have been updated

### Summary
Implements [SDL-0108](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0108-AutoCompleteList.md) to enable a list of autocomplete suggestions in the perform interaction keyboard.

##### Enhancements
* Keyboard used in Perform Interactions / the Screen Manager (searchable choice sets and keyboards) can now display a list of autocompletion suggestion options ([SDL-0108](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0108-AutoCompleteList.md)).

### Tasks Remaining:
- [x] Test against Core 6.0

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by NicoleYarroch** [comment]
This PR has been removed from Core 5.0.0 (and therefore 6.1.0). This PR will remain open for the time being, but is not targeted for a specific release.

File: SmartDeviceLink/SDLKeyboardProperties.h [hunk 1]
```diff
@@ -53,6 +76,13 @@ NS_ASSUME_NONNULL_BEGIN
  */
 @property (nullable, strong, nonatomic) NSString *autoCompleteText;
 
+/**
+ Allows an app to show a list of possible autocomplete suggestions as the user types
+
+ Optional, 1-100 items
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLKeyboardProperties.h]: Should add the max string length as well

File: SmartDeviceLink/SDLKeyboardDelegate.h [hunk 2]
```diff
@@ -17,10 +17,17 @@ NS_ASSUME_NONNULL_BEGIN
 /**
  This handler is called when you wish to update your autocomplete text in response to the user's input
 
- @param updatedAutocompleteText The new autocomplete text to use
+ @param updatedAutocompleteText The autocomplete results to use
  */
 typedef void(^SDLKeyboardAutocompleteCompletionHandler)(NSString  *_Nullable updatedAutocompleteText);
 
+/**
+ This handler is called when you wish to update your autocomplete text in response to the user's input.
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLKeyboardDelegate.h]: Should add that a max of 100 items is allowed to be shown at once

File: SmartDeviceLink/SDLKeyboardDelegate.h [hunk 3]
```diff
@@ -65,7 +72,9 @@ typedef void(^SDLKeyboardCharacterSetCompletionHandler)(NSArray<NSString *> *_Nu
  @param currentInputText The user's full current input text
  @param completionHandler A completion handler to update the autoCompleteText
  */
-- (void)updateAutocompleteWithInput:(NSString *)currentInputText completionHandler:(SDLKeyboardAutocompleteCompletionHandler)completionHandler;
+- (void)updateAutocompleteWithInput:(NSString *)currentInputText completionHandler:(SDLKeyboardAutocompleteCompletionHandler)completionHandler __deprecated_msg("Use updateAutocompleteWithInput:autoCompleteResultsHandler:");
+
+- (void)updateAutocompleteWithInput:(NSString *)currentInputText autoCompleteResultsHandler:(SDLKeyboardAutoCompleteResultsHandler)resultsHandler;
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLKeyboardDelegate.h]: Fixed

File: SmartDeviceLink/SDLKeyboardProperties.h (hunk 1 above)
Comment (joeljfischer) [reply SmartDeviceLink/SDLKeyboardProperties.h]: Fixed

File: SmartDeviceLink/SDLKeyboardDelegate.h (hunk 2 above)
Comment (joeljfischer) [reply SmartDeviceLink/SDLKeyboardDelegate.h]: Fixed

File: SmartDeviceLink/SDLPresentKeyboardOperation.m [hunk 4]
```diff
@@ -141,11 +141,20 @@ - (void)sdl_keyboardInputNotification:(SDLRPCNotificationNotification *)notifica
         [self.keyboardDelegate userDidSubmitInput:onKeyboard.data withEvent:onKeyboard.event];
     } else if ([onKeyboard.event isEqualToEnum:SDLKeyboardEventKeypress]) {
         // Notify of keypress
-        if ([self.keyboardDelegate respondsToSelector:@selector(updateAutocompleteWithInput:completionHandler:)]) {
-            [self.keyboardDelegate updateAutocompleteWithInput:onKeyboard.data completionHandler:^(NSString *updatedAutocompleteText) {
+        if ([self.keyboardDelegate respondsToSelector:@selector(updateAutocompleteWithInput:autoCompleteResultsHandler:)]) {
+            [self.keyboardDelegate updateAutocompleteWithInput:onKeyboard.data autoCompleteResultsHandler:^(NSArray<NSString *> * _Nullable updatedAutoCompleteList) {
+                weakself.keyboardProperties.autoCompleteList = (updatedAutoCompleteList.count > 0) ? updatedAutoCompleteList : nil;
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLPresentKeyboardOperation.m]: fixed

File: SmartDeviceLink/SDLPresentChoiceSetOperation.m [hunk 5]
```diff
@@ -211,11 +211,20 @@ - (void)sdl_keyboardInputNotification:(SDLRPCNotificationNotification *)notifica
         [self.keyboardDelegate userDidSubmitInput:onKeyboard.data withEvent:onKeyboard.event];
     } else if ([onKeyboard.event isEqualToEnum:SDLKeyboardEventKeypress]) {
         // Notify of keypress
-        if ([self.keyboardDelegate respondsToSelector:@selector(updateAutocompleteWithInput:completionHandler:)]) {
-            [self.keyboardDelegate updateAutocompleteWithInput:onKeyboard.data completionHandler:^(NSString *updatedAutocompleteText) {
+        if ([self.keyboardDelegate respondsToSelector:@selector(updateAutocompleteWithInput:autoCompleteResultsHandler:)]) {
+            [self.keyboardDelegate updateAutocompleteWithInput:onKeyboard.data autoCompleteResultsHandler:^(NSArray<NSString *> * _Nullable updatedAutoCompleteList) {
+                weakself.keyboardProperties.autoCompleteList = (updatedAutoCompleteList.count > 0) ? updatedAutoCompleteList : nil;
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLPresentChoiceSetOperation.m]: Fixed

File: SmartDeviceLink/SDLPresentChoiceSetOperation.m [hunk 6]
```diff
@@ -213,11 +213,27 @@ - (void)sdl_keyboardInputNotification:(SDLRPCNotificationNotification *)notifica
...
+        if ([self.keyboardDelegate respondsToSelector:@selector(updateAutocompleteWithInput:autoCompleteResultsHandler:)]) {
+            [self.keyboardDelegate updateAutocompleteWithInput:onKeyboard.data autoCompleteResultsHandler:^(NSArray<NSString *> * _Nullable updatedAutoCompleteList) {
+                NSArray<NSString *> *newList = nil;
+                if (updatedAutoCompleteList.count > 100) {
+                    newList = [updatedAutoCompleteList subarrayWithRange:NSMakeRange(0, 100)];
+                } else {
+                    newList = updatedAutoCompleteList;
+                }
+
+                weakself.keyboardProperties.autoCompleteList = (newList.count > 0) ? newList : nil;
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLPresentChoiceSetOperation.m]: `nil` needs to be `@[]`

File: SmartDeviceLink/SDLPresentKeyboardOperation.m [hunk 7]
```diff
@@ -143,11 +143,27 @@ - (void)sdl_keyboardInputNotification:(SDLRPCNotificationNotification *)notifica
...
+        if ([self.keyboardDelegate respondsToSelector:@selector(updateAutocompleteWithInput:autoCompleteResultsHandler:)]) {
+            [self.keyboardDelegate updateAutocompleteWithInput:onKeyboard.data autoCompleteResultsHandler:^(NSArray<NSString *> * _Nullable updatedAutoCompleteList) {
+                NSArray<NSString *> *newList = nil;
+                if (updatedAutoCompleteList.count > 100) {
+                    newList = [updatedAutoCompleteList subarrayWithRange:NSMakeRange(0, 100)];
+                } else {
+                    newList = updatedAutoCompleteList;
+                }
+
+                weakself.keyboardProperties.autoCompleteList = (newList.count > 0) ? newList : nil;
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLPresentKeyboardOperation.m]: `nil` needs to be `@[]`

File: SmartDeviceLinkTests/DevAPISpecs/SDLPresentChoiceSetOperationSpec.m [hunk 8]
```diff
@@ -234,7 +234,7 @@
                 NSString *inputData = @"Test";
                 SDLRPCNotificationNotification *notification = nil;
 
-                OCMStub([testKeyboardDelegate updateAutocompleteWithInput:[OCMArg any] completionHandler:([OCMArg invokeBlockWithArgs:inputData, nil])]);
+                OCMStub([testKeyboardDelegate updateAutocompleteWithInput:[OCMArg any] completionHandler:([OCMArg invokeBlockWithArgs:@[inputData], nil])]);
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/DevAPISpecs/SDLPresentChoiceSetOperationSpec.m]: Getting 3 failed test cases

File: SmartDeviceLinkTests/DevAPISpecs/SDLPresentKeyboardOperationSpec.m [hunk 9]
```diff
@@ -163,7 +163,7 @@
                 NSString *inputData = @"Test";
                 SDLRPCNotificationNotification *notification = nil;
 
-                OCMStub([testDelegate updateAutocompleteWithInput:[OCMArg any] completionHandler:([OCMArg invokeBlockWithArgs:inputData, nil])]);
+                OCMStub([testDelegate updateAutocompleteWithInput:[OCMArg any] completionHandler:([OCMArg invokeBlockWithArgs:@[inputData], nil])]);
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/DevAPISpecs/SDLPresentKeyboardOperationSpec.m]: Getting 3 failed test cases

File: SmartDeviceLink/SDLPresentChoiceSetOperation.m (hunk 6 above)
Comment (joeljfischer) [reply SmartDeviceLink/SDLPresentChoiceSetOperation.m]: Fixed

Comment (joeljfischer) [reply SmartDeviceLink/SDLPresentChoiceSetOperation.m]: Fixed

File: SmartDeviceLink/SDLPresentKeyboardOperation.m (hunk 7 above)
Comment (joeljfischer) [reply SmartDeviceLink/SDLPresentKeyboardOperation.m]: Fixed

File: SmartDeviceLinkTests/DevAPISpecs/SDLPresentChoiceSetOperationSpec.m (hunk 8 above)
Comment (joeljfischer) [reply SmartDeviceLinkTests/DevAPISpecs/SDLPresentChoiceSetOperationSpec.m]: Fixed

File: SmartDeviceLinkTests/DevAPISpecs/SDLPresentKeyboardOperationSpec.m (hunk 9 above)
Comment (joeljfischer) [reply SmartDeviceLinkTests/DevAPISpecs/SDLPresentKeyboardOperationSpec.m]: Fixed

**Pull request merged**